### PR TITLE
Atualização da dependência g4f e adição de suporte ao Docker

### PR DESCRIPTION
**Pull request by clouraLabs**
## Type
enhancement, configuration changes

___
## Description
- Atualizada a dependência g4f para a versão 0.1.9.2 no arquivo pyproject.toml para melhorar a performance e segurança da aplicação.
- Adicionado um Dockerfile para a contenerização da aplicação, que inclui:
  - Uso da imagem base python:3.11-slim-buster.
  - Definição do diretório de trabalho para /app.
  - Instalação do git e Poetry.
  - Cópia dos conteúdos do diretório atual para o contêiner.
  - Instalação das dependências com o Poetry sem instalar o pacote em si.
  - Exposição da porta 7001 para a aplicação.
  - Configuração do comando para executar main.py usando o Poetry.

___
## PR changes walkthrough
<table><thead><tr><th></th><th>Relevant files&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>Dockerfile&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        Dockerfile<br><br>
        <strong>Adicionado Dockerfile para contenerização da aplicação, <br>incluindo configuração do ambiente, instalação de <br>dependências e definição do comando de execução.</strong>
      </ul>
    </details>
  </td>
  <td><a href="https://github.com/tgscan-dev/g4f-api/pull/2/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557"> +26/-0</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        pyproject.toml<br><br>
        <strong>Atualizada a versão da dependência g4f para 0.1.9.2.</strong>
      </ul>
    </details>
  </td>
  <td><a href="https://github.com/tgscan-dev/g4f-api/pull/2/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711"> +1/-1</a></td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

### CODE REVIEW

**Comment by clouraen** [comment]
Preparing review...

**Comment by clouraLabs** [comment]
## PR Analysis

- 🎯 **Main theme:** Dependency updates and Dockerization
- 📝 **PR summary:** This PR updates the g4f package and other dependencies to their latest versions to improve performance and security. It also adds a Dockerfile to the project, allowing the application to be containerized for easier deployment and running in different environments.
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** No
- ⏱️ **Estimated effort to review [1-5]:** 2, because the changes are straightforward and involve updating dependencies and adding a Dockerfile. However, the impact of the dependency updates needs to be carefully reviewed.
- 🔒 **Security concerns:** No
## PR Feedback

- 💡 **General suggestions:** It would be beneficial to include a .dockerignore file to prevent unnecessary files from being included in the Docker image. Also, consider pinning the versions of the dependencies in the Dockerfile to ensure consistent builds.


- <details><summary> 🤖 Code feedback:</summary>


  - **relevant file:** ``Dockerfile``  
   **suggestion:** Consider using a multi-stage build to reduce the final image size. This can be done by installing dependencies and running tests in an intermediate image, and then copying only the necessary files into the final image. [important]  
   **relevant line:** [FROM python:3.11-slim-buster](https://github.com/tgscan-dev/g4f-api/pull/2/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R2)  


  - **relevant file:** ``Dockerfile``  
   **suggestion:** It's a good practice to run the application with a non-root user in the Docker container for security reasons. You can create a non-root user and switch to it before running the application. [important]  
   **relevant line:** [CMD ["poetry", "run", "python3", "main.py"]](https://github.com/tgscan-dev/g4f-api/pull/2/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R26)  


  - **relevant file:** ``pyproject.toml``  
   **suggestion:** Consider using the caret (^) version specifier for the g4f package to allow updates that do not modify the left-most non-zero digit in the version. This can help to receive patches and minor updates automatically. [medium]  
   **relevant line:** [g4f = "0.1.9.2"](https://github.com/tgscan-dev/g4f-api/pull/2/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R14)  

</details>

**Comment by clouraen** [comment]
Preparing PR description...

**Comment by clouraen** [comment]
**[PR Description](https://github.com/tgscan-dev/g4f-api/pull/2)** updated to latest commit (https://github.com/tgscan-dev/g4f-api/commit/76ded30d01373d4e8c6b6fd797ff7e0965be637b)

**Comment by clouraen** [comment]
Preparing PR description...

**Comment by clouraen** [comment]
**[PR Description](https://github.com/tgscan-dev/g4f-api/pull/2)** updated to latest commit (https://github.com/tgscan-dev/g4f-api/commit/76ded30d01373d4e8c6b6fd797ff7e0965be637b)

**Comment by clouraen** [comment]
Preparing PR description...

**Comment by clouraen** [comment]
**[PR Description](https://github.com/tgscan-dev/g4f-api/pull/2)** updated to latest commit (https://github.com/tgscan-dev/g4f-api/commit/76ded30d01373d4e8c6b6fd797ff7e0965be637b)

**Comment by clouraen** [comment]
Preparing PR description...

**Comment by clouraen** [comment]
**[PR Description](https://github.com/tgscan-dev/g4f-api/pull/2)** updated to latest commit (https://github.com/tgscan-dev/g4f-api/commit/76ded30d01373d4e8c6b6fd797ff7e0965be637b)

**Comment by clouraen** [comment]
Preparing suggestions...

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Use an official Python runtime as a parent image
+FROM python:3.11-slim-buster
+
+# Set the working directory in the container to /app
+WORKDIR /app
+
+# Install git
+RUN apt-get update && \
+    apt-get upgrade -y
```
Comment (clouraLabs) [inline Dockerfile]: **Suggestion:** Consider cleaning up the apt-get cache after installing packages to reduce the Docker image size.
```suggestion
RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+# Install Poetry
+RUN pip install poetry
```
Comment (clouraLabs) [inline Dockerfile]: **Suggestion:** Pin the versions of the packages you are installing to ensure reproducible builds.
```suggestion
RUN pip install poetry==1.1.13  # Replace with the desired version
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# Install dependencies using poetry
+# The --no-root option is used to avoid installing the package (defined in pyproject.toml) itself
+# The --no-interaction option is used to avoid interactive prompts
+RUN poetry install --no-root --no-interaction
+
+# Expose port 7001 for the application
+EXPOSE 7001
+
+# Run main.py when the container launches
+CMD ["poetry", "run", "python3", "main.py"]
```
Comment (clouraLabs) [inline Dockerfile]: **Suggestion:** Use a multi-stage build to reduce the final image size by not including unnecessary build dependencies.
```suggestion
# Build stage
FROM python:3.11-slim-buster as builder
WORKDIR /app
RUN apt-get update && \
    apt-get install -y --no-install-recommends git && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
RUN pip install poetry==1.1.13
COPY . /app
RUN poetry install --no-root --no-interaction

# Final stage
FROM python:3.11-slim-buster
WORKDIR /app
COPY --from=builder /app /app
EXPOSE 7001
CMD ["poetry", "run", "python3", "main.py"]
```

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+# Install git
+RUN apt-get update && \
+    apt-get upgrade -y
+
+# Install Poetry
+RUN pip install poetry
+
+# Copy the current directory contents into the container at /app
+COPY . /app
```
Comment (clouraLabs) [inline Dockerfile]: **Suggestion:** Use COPY with more granularity to leverage Docker cache layers effectively.
```suggestion
COPY pyproject.toml poetry.lock* /app/
COPY src /app/src
# Continue copying other necessary files and directories individually
```

**Comment by clouraen** [comment]
Generating Documentation...

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# Use an official Python runtime as a parent image
```
Comment (clouraLabs) [inline Dockerfile:1]: **Suggestion:** Proposed documentation
```suggestion
# Dockerfile para a contenerização da aplicação.
#
# Este Dockerfile descreve os passos para criar uma imagem Docker que
# contém a aplicação Python. A imagem base é python:3.11-slim-buster,
# que é uma imagem leve e adequada para aplicações Python.
#
# As etapas incluem:
# - Definir /app como diretório de trabalho.
# - Instalar o git.
# - Instalar a ferramenta de gerenciamento de dependências Poetry.
# - Copiar os arquivos necessários para o contêiner.
# - Instalar as dependências do projeto usando o Poetry.
# - Expor a porta 7001 para a aplicação.
# - Definir o comando padrão para executar a aplicação usando o Poetry.
# Use an official Python runtime as a parent image
```

**Pull request merged**